File: src/main/java/org/veupathdb/lib/gradle/container/util/StackIterator.java
package org.veupathdb.lib.gradle.container.util;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Comparator;
import java.util.Iterator;
import java.util.Spliterator;
import java.util.Stack;
import java.util.function.Consumer;

/**
 * Iterator Wrapping a {@link Stack} instance.
 *
 * @param <T> Generic type of this iterator and of the wrapped {@code Stack}
 *            instance.
 *
 * @since 1.1.0
 */
public record StackIterator<T>(
  @NotNull Stack<T> stack) implements Iterable<T>, Iterator<T>, Spliterator<T> {
  public StackIterator(@NotNull Stack<T> stack) {
    this.stack = stack;
  }

  ////  Iterable  ////////////////////////////////////////////////////////////////////////////////

  @Override
  @NotNull
  public Iterator<T> iterator() {
    return this;
  }

  @Override
  @NotNull
  public Spliterator<T> spliterator() {
    return this;
  }

  @Override
  public void forEach(@NotNull final Consumer<? super T> action) {
    while (!stack.isEmpty()) {
      action.accept(stack.pop());
    }
  }

  ////  Iterator  ////////////////////////////////////////////////////////////////////////////////

  @Override
  public boolean hasNext() {
    return !stack.empty();
  }

  @Override
  @NotNull
  public T next() {
    return stack.pop();
  }

  @Override
  public void forEachRemaining(@NotNull final Consumer<? super T> action) {
    while (!stack.isEmpty()) {
      action.accept(stack.pop());
    }
  }

  @Override
  public void remove() {}

  ////  Spliterator  /////////////////////////////////////////////////////////////////////////////

  @Override
  public boolean tryAdvance(Consumer<? super T> action) {
    if (stack.isEmpty()) {
      return false;
    }

    action.accept(stack.pop());
    return true;
  }

  @Override
  @Nullable
  public Spliterator<T> trySplit() {
    return null;
  }

  @Override
  public long estimateSize() {
    return stack.size();
  }

  @Override
  public int characteristics() {
    return ORDERED | SORTED | DISTINCT | NONNULL;
  }

  @Override
  public Comparator<? super T> getComparator() {
    return null;
  }
}

File: src/main/java/org/veupathdb/lib/gradle/container/util/StackIterator.kt
package org.veupathdb.lib.gradle.container.util

import java.util.Spliterator
import java.util.Spliterator.*
import java.util.Stack
import java.util.function.Consumer


/**
 * Iterator Wrapping a [Stack] instance.
 *
 * @param T Generic type of this iterator and of the wrapped [Stack] instance.
 *
 * @since 1.1.0
 */
data class StackIterator<T>(val stack: Stack<T>)
  : Iterable<T>
  , Iterator<T>
  , Spliterator<T>
{

  ////  Iterable  ////////////////////////////////////////////////////////////////////////////////

  override fun iterator() = this

  override fun spliterator() = this

  override fun forEach(action: Consumer<in T>) {
    while (!stack.isEmpty()) {
      action.accept(stack.pop())
    }
  }

  ////  Iterator  ////////////////////////////////////////////////////////////////////////////////

  override fun hasNext() = !stack.empty()

  override fun next(): T = stack.pop()

  override fun forEachRemaining(action: Consumer<in T>) {
    while (!stack.isEmpty()) {
      action.accept(stack.pop())
    }
  }

  ////  Spliterator  /////////////////////////////////////////////////////////////////////////////

  override fun tryAdvance(action: Consumer<in T>): Boolean {
    if (stack.isEmpty()) {
      return false
    }

    action.accept(stack.pop())
    return true
  }

  override fun trySplit(): Spliterator<T>? {
    return null
  }

  override fun estimateSize(): Long {
    return stack.size.toLong()
  }

  override fun characteristics(): Int {
    return ORDERED or SORTED or DISTINCT or NONNULL
  }

  override fun getComparator(): Comparator<in T>? {
    return null
  }
}
